sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/PartitionBasedLoadBalancer.java

```diff
@@ -0,0 +1,352 @@
+                    || partitionOwnership.eTag() == null;
+            });
+    }
+
+    /*
+     * Find the event processor that owns the maximum number of partitions and steal a random partition
+     * from it.
+     */
+    private String findPartitionToSteal(final Map<String, List<PartitionOwnership>> ownerPartitionMap) {
+        Map.Entry<String, List<PartitionOwnership>> ownerWithMaxPartitions = ownerPartitionMap.entrySet()
+            .stream()
+            .max(Comparator.comparingInt(entry -> entry.getValue().size()))
+            .get();
+        int numberOfPartitions = ownerWithMaxPartitions.getValue().size();
+        logger.info("Owner id {} owns {} partitions, stealing a partition from it", ownerWithMaxPartitions.getKey(),
+            numberOfPartitions);
+        return ownerWithMaxPartitions.getValue().get(RANDOM.nextInt(numberOfPartitions)).partitionId();
+    }
+
+    /*
+     * When the load is balanced, all active event processors own at least {@code minPartitionsPerEventProcessor}
+     * and only {@code numberOfEventProcessorsWithAdditionalPartition} event processors will own 1 additional
+     * partition.
+     */
+    private boolean isLoadBalanced(final int minPartitionsPerEventProcessor,
+        final int numberOfEventProcessorsWithAdditionalPartition,
+        final Map<String, List<PartitionOwnership>> ownerPartitionMap) {
+
+        int count = 0;
+        for (List<PartitionOwnership> partitionOwnership : ownerPartitionMap.values()) {
+            int numberOfPartitions = partitionOwnership.size();
+            if (numberOfPartitions < minPartitionsPerEventProcessor
+                || numberOfPartitions > minPartitionsPerEventProcessor + 1) {
+                return false;
+            }
+
+            if (numberOfPartitions == minPartitionsPerEventProcessor + 1) {
+                count++;
+            }
+        }
+        return count == numberOfEventProcessorsWithAdditionalPartition;
+    }
+
+    /*
+     * This method is called after determining that the load is not balanced. This method will evaluate
+     * if the current event processor should own more partitions. Specifically, this method returns true if the
+     * current event processor owns less than the minimum number of partitions or if it owns the minimum number
+     * and no other event processor owns lesser number of partitions than this event processor.
+     */
+    private boolean shouldOwnMorePartitions(final int minPartitionsPerEventProcessor,
+        final Map<String, List<PartitionOwnership>> ownerPartitionMap) {
+
+        int numberOfPartitionsOwned = ownerPartitionMap.get(this.ownerId).size();
+
+        int leastPartitionsOwnedByAnyEventProcessor =
+            ownerPartitionMap.values().stream().min(Comparator.comparingInt(List::size)).get().size();
+
+        return numberOfPartitionsOwned < minPartitionsPerEventProcessor
+            || numberOfPartitionsOwned == leastPartitionsOwnedByAnyEventProcessor;
```

Comment:
I'm not sure if this second condition makes sense.

Scenario:
There are 6 partitions total, and 4 EventProcessors.
The expected distribution would be 2 EventProcessors owning 2 partitions, and 2 EventProcessors owning 1 partition.

Let's say our EventProcessor currently owns 1 partition. In this case, `numberOfPartitionsOwned === leastPartitionsOwnedByAnyEventProcessor` will be true, indicating that the EventProcessor should own more partitions. This would be incorrect though since based on the scenario, we know we've reached an optimal load balance.

I think the EventProcessor should own more partitions if:
1. numberOfPartitionsOwned is less than the minimum partitions per event processor.
2. The sum of partitions actively owned by any processor is less than the total number of partitions __AND__ the numberOfPartitionsOwned is less than the `minPartitionsPerEventProcessor + 1`.

Reply:
We won't reach this code if the load is already balanced. On line 189, there's a check to see if the load is already balanced. This method gets called only when the load is not balanced.